GlobalRefProxy.c

```diff
@@ -0,0 +1,70 @@
+#include <nspr.h>
+#include <nss.h>
+#include <jni.h>
+
+#include "java_ids.h"
+#include "jssutil.h"
+
+jobject
+JSS_PR_wrapGlobalRef(JNIEnv *env, jobject *ref)
+{
+    jbyteArray pointer = NULL;
+    jclass proxyClass;
+    jmethodID constructor;
+    jobject refObj = NULL;
+
+    PR_ASSERT(env != NULL && ref != NULL && *ref != NULL);
+
+    /* Convert pointer to byte array. */
+    pointer = JSS_ptrToByteArray(env, *ref);
+
+    /* Lookup the class and constructor. */
+    proxyClass = (*env)->FindClass(env, GLOBAL_REF_PROXY_CLASS_NAME);
+    if (proxyClass == NULL) {
+        ASSERT_OUTOFMEM(env);
+        goto finish;
+    }
+
+    constructor = (*env)->GetMethodID(env, proxyClass,
+                                      PLAIN_CONSTRUCTOR,
+                                      GLOBAL_REF_CONSTRUCTOR_SIG);
+    if (constructor == NULL) {
+        ASSERT_OUTOFMEM(env);
+        goto finish;
+    }
+
+    /* Call the constructor. */
+    refObj = (*env)->NewObject(env, proxyClass, constructor, pointer);
+
+finish:
+    if (refObj == NULL && *ref != NULL) {
```

Comment:
Is `*ref != NULL` still needed? There's an assertion for it in line 16.

Reply:
"eh" -- `PR_ASSERT` behaves like `assert(...)` in C (which only triggers when `DEBUG` is defined; we don't define it in release modes). I think `PR_ASSERT` triggers when `DEBUG` or `PR_DEBUG` is defined (the latter we don't define anywhere). This is similar to `assert(...)` in Java, where it only gets triggered when `-ea` is passed to the running JVM. Otherwise, they'll be ignored.

Most everything (in NSS or JSS) handles a `NULL` free just fine. I'm not sure if `DeleteGlobalRef` (which comes from the JVM) does. It isn't clear from [the docs](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html) if it handles `NULL` safely.

Reply:
Ah, OK. I'm wondering if we should continue to use PR_ASSERTs or convert them into real code. Regardless, feel free to push.

Reply:
Yeah, at some point in the future, I'd migrate to `PR_ASSERT`'s (so we get a nice crash we can attach to GDB with when built in debug mode) and `if` statements with proper exceptions. Some of these we don't need to assert though (`env != NULL` is a really weird edge case).